<!DOCTYPE html>
<html lang="en">
 <head>
  <meta charset="utf-8">
  <title>PHPUnit マニュアル &#8211; 第2章 PHPUnit 用のテストの書き方</title>
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  <link rel="canonical" href="writing-tests-for-phpunit.html">
  <link href="css/bootstrap.min.css" rel="stylesheet">
  <link href="css/highlight.css" rel="stylesheet">
  <link href="css/style.css" rel="stylesheet">
  <link href='https://fonts.googleapis.com/css?family=Open+Sans:400,700%7CSource+Code+Pro:400,700' rel='stylesheet' type='text/css'>
  <!--[if lt IE 9]><script src="js/html5shiv.min.js"></script><![endif]-->
 </head>
 <body>
  
  <div class="container-fluid">
   <div class="row">
    <div style='padding: 2.0em'>
     <div class="row">
      <div class="col-md-1 pull-left prev-nav"><a accesskey="p" href="installation.html">戻る</a></div>
      <div class="col-md-1 pull-right next-nav"><a accesskey="n" href="textui.html">次へ</a></div>
     </div>
<div class="chapter"><div class="titlepage"><div><div><h1 class="title"><a id="writing-tests-for-phpunit"></a>第2章 PHPUnit 用のテストの書き方</h1></div></div></div><p>
    <a id="idp139120" class="indexterm"></a>

    <a class="xref" href="writing-tests-for-phpunit.html#writing-tests-for-phpunit.examples.StackTest.php" title="例 2.1: PHPUnit での配列操作のテスト">例 2.1</a> で、
    PHP の配列操作のテストを PHPUnit 用に書く方法を示します。
    この例では、PHPUnit を使ったテストを書く際の基本的な決まり事や手順を紹介します。
  </p><div class="orderedlist"><ol class="orderedlist" type="1"><li class="listitem"><p><code class="literal">Class</code> という名前のクラスのテストは、<code class="literal">ClassTest</code> という名前のクラスに記述します。</p></li><li class="listitem"><p><code class="literal">ClassTest</code> は、(ほとんどの場合) <code class="literal">PHPUnit_Framework_TestCase</code> を継承します。</p></li><li class="listitem"><p>テストは、<code class="literal">test*</code> という名前のパブリックメソッドとなります。</p><p><a id="idp119184" class="indexterm"></a><a id="idp119600" class="indexterm"></a>あるいは、<code class="literal">@test</code> アノテーションをメソッドのコメント部で使用することで、それがテストメソッドであることを示すこともできます。</p></li><li class="listitem"><p>テストメソッドの中で <code class="literal">assertEquals()</code> のようなアサーションメソッド (<a class="xref" href="appendixes.assertions.html" title="付録A アサーション">付録 A</a> を参照ください) を使用して、期待される値と実際の値が等しいことを確かめます。</p></li></ol></div><div class="example"><a id="writing-tests-for-phpunit.examples.StackTest.php"></a><p class="title"><strong>例 2.1: PHPUnit での配列操作のテスト</strong></p><div class="example-contents"><pre class="programlisting">&lt;?php
class StackTest extends PHPUnit_Framework_TestCase
{
    public function testPushAndPop()
    {
        $stack = array();
        $this-&gt;assertEquals(0, count($stack));

        array_push($stack, 'foo');
        $this-&gt;assertEquals('foo', $stack[count($stack)-1]);
        $this-&gt;assertEquals(1, count($stack));

        $this-&gt;assertEquals('foo', array_pop($stack));
        $this-&gt;assertEquals(0, count($stack));
    }
}
?&gt;</pre></div></div><br class="example-break"></br><div class="blockquote"><table border="0" class="blockquote" style="width: 100%; cellspacing: 0; cellpadding: 0;" summary="Block quote"><tr><td width="10%" valign="top"> </td><td width="80%" valign="top"><p>
      Whenever you are tempted to type something into a
      <code class="literal">print</code> statement or a debugger expression, write it
      as a test instead.
    </p><p>
      何かを <code class="literal">print</code> 文やデバッガの式に書きたくなったときは、
      代わりにその内容をテストに書くようにするんだ。
    </p></td><td width="10%" valign="top"> </td></tr><tr><td width="10%" valign="top"> </td><td colspan="2" align="right" valign="top">--<span class="attribution">Martin Fowler</span></td></tr></table></div><div class="section"><div class="titlepage"><div><div><h2 class="title" style="clear: both"><a id="writing-tests-for-phpunit.test-dependencies"></a>テストの依存性</h2></div></div></div><div class="blockquote"><table border="0" class="blockquote" style="width: 100%; cellspacing: 0; cellpadding: 0;" summary="Block quote"><tr><td width="10%" valign="top"> </td><td width="80%" valign="top"><p>
        Unit Tests are primarily written as a good practice to help developers
        identify and fix bugs, to refactor code and to serve as documentation
        for a unit of software under test. To achieve these benefits, unit tests
        ideally should cover all the possible paths in a program. One unit test
        usually covers one specific path in one function or method. However a
        test method is not necessary an encapsulated, independent entity. Often
        there are implicit dependencies between test methods, hidden in the
        implementation scenario of a test.
      </p><p>
        ユニットテストを書くそもそもの目的は、バグの発見と修正や
        コードのリファクタリングを開発者がやりやすくすること。
        そしてテスト対象のソフトウェアのドキュメントとしての役割を果たすことだ。
        これらの目的を達成するためには、
        ユニットテストがプログラム内のすべてのルートをカバーしていることが理想である。
        ひとつのユニットテストがカバーするのは、
        通常はひとつの関数やメソッド内の特定のルートだけとなる。
        しかし、テストメソッドは必ずしもカプセル化して独立させる必要はない。
        複数のテストメソッドの間に暗黙の依存性があって、
        隠された実装シナリオがテストの中にあるのもよくあることだ。
      </p></td><td width="10%" valign="top"> </td></tr><tr><td width="10%" valign="top"> </td><td colspan="2" align="right" valign="top">--<span class="attribution">Adrian Kuhn et. al.</span></td></tr></table></div><p>
      <a id="idp218464" class="indexterm"></a>

      PHPUnit は、テストメソッド間の依存性の明示的な宣言をサポートしています。
      この依存性とは、テストメソッドが実行される順序を定義するものではありません。
      プロデューサーがテストフィクスチャを作ってそのインスタンスを返し、
      依存するコンシューマーがそれを受け取って利用するというものです。
    </p><div class="itemizedlist"><ul class="itemizedlist" style="list-style-type: disc; "><li class="listitem"><p>プロデューサーとは、返り値としてテスト対象のユニットを生成するテストメソッドのこと。</p></li><li class="listitem"><p>コンシューマーとは、プロデューサーの返り値に依存するテストメソッドのこと。</p></li></ul></div><p>
      <a id="idp221728" class="indexterm"></a>
      <a id="idp222304" class="indexterm"></a>

      <a class="xref" href="writing-tests-for-phpunit.html#writing-tests-for-phpunit.examples.StackTest2.php" title="例 2.2: @depends アノテーションを使った依存性の表現">例 2.2</a>
      は、<code class="literal">@depends</code> アノテーションを使ってテストメソッドの依存性をあらわす例です。
    </p><div class="example"><a id="writing-tests-for-phpunit.examples.StackTest2.php"></a><p class="title"><strong>例 2.2: <code class="literal">@depends</code> アノテーションを使った依存性の表現</strong></p><div class="example-contents"><pre class="programlisting">&lt;?php
class StackTest extends PHPUnit_Framework_TestCase
{
    public function testEmpty()
    {
        $stack = array();
        $this-&gt;assertEmpty($stack);

        return $stack;
    }

    /**
     * @depends testEmpty
     */
    public function testPush(array $stack)
    {
        array_push($stack, 'foo');
        $this-&gt;assertEquals('foo', $stack[count($stack)-1]);
        $this-&gt;assertNotEmpty($stack);

        return $stack;
    }

    /**
     * @depends testPush
     */
    public function testPop(array $stack)
    {
        $this-&gt;assertEquals('foo', array_pop($stack));
        $this-&gt;assertEmpty($stack);
    }
}
?&gt;</pre></div></div><br class="example-break"></br><p>
      上の例では、まず最初のテスト <code class="literal">testEmpty()</code>
      で新しい配列を作り、それが空であることを確かめます。
      このテストは、フィクスチャを返します。
      二番目のテスト <code class="literal">testPush()</code> は
      <code class="literal">testEmpty()</code> に依存しており、
      依存するテストの結果を引数として受け取ります。
      最後の <code class="literal">testPop()</code> は
      <code class="literal">testPush()</code> に依存しています。
    </p><div class="alert alert-info"><h3 class="title">注記</h3><p>
        <a id="idp70192" class="indexterm"></a>
        <a id="idp70768" class="indexterm"></a>

        プロデューサーの生成する戻り値は、デフォルトでは「そのままの形式」でコンシューマーに渡されます。
        つまり、プロデューサーがオブジェクトを戻した場合は、そのオブジェクトへの参照がコンシューマーに渡されるということです。
        参照ではなくオブジェクトのコピーを渡す必要がある場合は、<code class="code">@depends</code>
        ではなく <code class="code">@depends clone</code> を使う必要があります。
      </p></div><p>
      <a id="idp73296" class="indexterm"></a>

      問題の局所化を手早く行うには、失敗したテストに目を向けやすくしたいものです。
      そのため PHPUnit では、
      あるテストが失敗したときにはそのテストに依存する他のテストの実行をスキップします。
      テスト間の依存性を活用して問題点を見つけやすくしている例を
      <a class="xref" href="writing-tests-for-phpunit.html#writing-tests-for-phpunit.examples.DependencyFailureTest.php" title="例 2.3: テストの依存性の活用">例 2.3</a>
      に示します。
    </p><div class="example"><a id="writing-tests-for-phpunit.examples.DependencyFailureTest.php"></a><p class="title"><strong>例 2.3: テストの依存性の活用</strong></p><div class="example-contents"><pre class="programlisting">&lt;?php
class DependencyFailureTest extends PHPUnit_Framework_TestCase
{
    public function testOne()
    {
        $this-&gt;assertTrue(FALSE);
    }

    /**
     * @depends testOne
     */
    public function testTwo()
    {
    }
}
?&gt;</pre><pre class="screen"><strong class="userinput"><code>phpunit --verbose DependencyFailureTest</code></strong>
PHPUnit 5.3.0 by Sebastian Bergmann and contributors.

FS

Time: 0 seconds, Memory: 5.00Mb

There was 1 failure:

1) DependencyFailureTest::testOne
Failed asserting that false is true.

/home/sb/DependencyFailureTest.php:6

There was 1 skipped test:

1) DependencyFailureTest::testTwo
This test depends on "DependencyFailureTest::testOne" to pass.


FAILURES!
Tests: 1, Assertions: 1, Failures: 1, Skipped: 1.</pre></div></div><br class="example-break"></br><p>
      ひとつのテストに複数の <code class="literal">@depends</code> アノテーションをつけることもできます。
      PHPUnit はテストが実行される順序を変更しないので、
      テストが実行されるときに確実に依存性が満たされているようにしておく必要があります。
    </p><p>
      複数の <code class="literal">@depends</code> アノテーションを持つテストは、
      最初のプロデューサーからのフィクスチャを最初の引数、二番目のプロデューサーからのフィクスチャを二番目の引数、……
      として受け取ります。
      <a class="xref" href="writing-tests-for-phpunit.html#writing-tests-for-phpunit.examples.MultipleDependencies.php" title="例 2.4: 複数の依存性を持つテスト">例 2.4</a>
      を参照ください。
    </p><div class="example"><a id="writing-tests-for-phpunit.examples.MultipleDependencies.php"></a><p class="title"><strong>例 2.4: 複数の依存性を持つテスト</strong></p><div class="example-contents"><pre class="programlisting">&lt;?php
class MultipleDependenciesTest extends PHPUnit_Framework_TestCase
{
    public function testProducerFirst()
    {
        $this-&gt;assertTrue(true);
        return 'first';
    }

    public function testProducerSecond()
    {
        $this-&gt;assertTrue(true);
        return 'second';
    }

    /**
     * @depends testProducerFirst
     * @depends testProducerSecond
     */
    public function testConsumer()
    {
        $this-&gt;assertEquals(
            array('first', 'second'),
            func_get_args()
        );
    }
}
?&gt;</pre><pre class="screen"><strong class="userinput"><code>phpunit --verbose MultipleDependenciesTest</code></strong>
PHPUnit 5.3.0 by Sebastian Bergmann and contributors.

...

Time: 0 seconds, Memory: 3.25Mb

OK (3 tests, 3 assertions)</pre></div></div><br class="example-break"></br></div><div class="section"><div class="titlepage"><div><div><h2 class="title" style="clear: both"><a id="writing-tests-for-phpunit.data-providers"></a>データプロバイダ</h2></div></div></div><p>
      <a id="idp329712" class="indexterm"></a>
      <a id="idp330288" class="indexterm"></a>
      テストメソッドには任意の引数を渡すことができます。
      この引数は、データプロバイダメソッド
      (<a class="xref" href="writing-tests-for-phpunit.html#writing-tests-for-phpunit.data-providers.examples.DataTest.php" title="例 2.5: 配列の配列を返すデータプロバイダの使用">例 2.5</a>
      の <code class="literal">additionProvider()</code>)
      で指定します。使用するデータプロバイダメソッドを指定するには
      <code class="literal">@dataProvider</code> アノテーションを使用します。
    </p><p>
      データプロバイダメソッドは、<code class="literal">public</code>
      でなければなりません。また、
      メソッドの返り値の型は、配列の配列あるいはオブジェクト
      (<code class="literal">Iterator</code> インターフェイスを実装しており、
      反復処理の際に配列を返すもの) である必要があります。
      この返り値の各要素に対して、その配列の中身を引数としてテストメソッドがコールされます。
    </p><div class="example"><a id="writing-tests-for-phpunit.data-providers.examples.DataTest.php"></a><p class="title"><strong>例 2.5: 配列の配列を返すデータプロバイダの使用</strong></p><div class="example-contents"><pre class="programlisting">&lt;?php
class DataTest extends PHPUnit_Framework_TestCase
{
    /**
     * @dataProvider additionProvider
     */
    public function testAdd($a, $b, $expected)
    {
        $this-&gt;assertEquals($expected, $a + $b);
    }

    public function additionProvider()
    {
        return array(
          array(0, 0, 0),
          array(0, 1, 1),
          array(1, 0, 1),
          array(1, 1, 3)
        );
    }
}
?&gt;</pre><pre class="screen"><strong class="userinput"><code>phpunit DataTest</code></strong>
PHPUnit 5.3.0 by Sebastian Bergmann and contributors.

...F

Time: 0 seconds, Memory: 5.75Mb

There was 1 failure:

1) DataTest::testAdd with data set #3 (1, 1, 3)
Failed asserting that 2 matches expected 3.

/home/sb/DataTest.php:9

FAILURES!
Tests: 4, Assertions: 4, Failures: 1.</pre></div></div><br class="example-break"></br><p>
      大量のデータセットを使う場合は、デフォルトの数字を使うのではなく、各データセットに文字列の名前をつけておくと便利です。
      出力もよりわかりやすくなり、テストを失敗させたデータセットの名前もわかるようになります。
    </p><div class="example"><a id="writing-tests-for-phpunit.data-providers.examples.DataTest1.php"></a><p class="title"><strong>例 2.6: データプロバイダでの名前つきデータセットの使用</strong></p><div class="example-contents"><pre class="programlisting">&lt;?php
class DataTest extends PHPUnit_Framework_TestCase
{
    /**
     * @dataProvider additionProvider
     */
    public function testAdd($a, $b, $expected)
    {
        $this-&gt;assertEquals($expected, $a + $b);
    }

    public function additionProvider()
    {
        return array(
          'adding zeros' =&gt; array(0, 0, 0),
          'zero plus one' =&gt; array(0, 1, 1),
          'one plus zero' =&gt; array(1, 0, 1),
          'one plus one' =&gt; array(1, 1, 3)
        );
    }
}
?&gt;</pre><pre class="screen"><strong class="userinput"><code>phpunit DataTest</code></strong>
PHPUnit 4.6.0 by Sebastian Bergmann and contributors.

...F

Time: 0 seconds, Memory: 5.75Mb

There was 1 failure:

1) DataTest::testAdd with data set "one plus one" (1, 1, 3)
Failed asserting that 2 matches expected 3.

/home/sb/DataTest.php:9

FAILURES!
Tests: 4, Assertions: 4, Failures: 1.</pre></div></div><br class="example-break"></br><div class="example"><a id="writing-tests-for-phpunit.data-providers.examples.DataTest2.php"></a><p class="title"><strong>例 2.7: Iterator オブジェクトを返すデータプロバイダの使用</strong></p><div class="example-contents"><pre class="programlisting">&lt;?php
require 'CsvFileIterator.php';

class DataTest extends PHPUnit_Framework_TestCase
{
    /**
     * @dataProvider additionProvider
     */
    public function testAdd($a, $b, $expected)
    {
        $this-&gt;assertEquals($expected, $a + $b);
    }

    public function additionProvider()
    {
        return new CsvFileIterator('data.csv');
    }
}
?&gt;</pre><pre class="screen"><strong class="userinput"><code>phpunit DataTest</code></strong>
PHPUnit 5.3.0 by Sebastian Bergmann and contributors.

...F

Time: 0 seconds, Memory: 5.75Mb

There was 1 failure:

1) DataTest::testAdd with data set #3 ('1', '1', '3')
Failed asserting that 2 matches expected '3'.

/home/sb/DataTest.php:11

FAILURES!
Tests: 4, Assertions: 4, Failures: 1.</pre></div></div><br class="example-break"></br><div class="example"><a id="writing-tests-for-phpunit.data-providers.examples.CsvFileIterator.php"></a><p class="title"><strong>例 2.8: CsvFileIterator クラス</strong></p><div class="example-contents"><pre class="programlisting">&lt;?php
class CsvFileIterator implements Iterator {
    protected $file;
    protected $key = 0;
    protected $current;

    public function __construct($file) {
        $this-&gt;file = fopen($file, 'r');
    }

    public function __destruct() {
        fclose($this-&gt;file);
    }

    public function rewind() {
        rewind($this-&gt;file);
        $this-&gt;current = fgetcsv($this-&gt;file);
        $this-&gt;key = 0;
    }

    public function valid() {
        return !feof($this-&gt;file);
    }

    public function key() {
        return $this-&gt;key;
    }

    public function current() {
        return $this-&gt;current;
    }

    public function next() {
        $this-&gt;current = fgetcsv($this-&gt;file);
        $this-&gt;key++;
    }
}
?&gt;</pre></div></div><br class="example-break"></br><p>
      <a id="idp347360" class="indexterm"></a>
      <a id="idp347936" class="indexterm"></a>
      <a id="idp348512" class="indexterm"></a>

      <code class="literal">@dataProvider</code> で指定したメソッドと
      <code class="literal">@depends</code> で指定したテストの両方からの入力を受け取るテストの場合、
      データプロバイダからの引数のほうが依存するテストからの引数より先にきます。
      依存するテストからの引数は、どちらのデータセットに対しても同じになります。
      <a class="xref" href="writing-tests-for-phpunit.html#writing-tests-for-phpunit.data-providers.examples.DependencyAndDataProviderCombo.php" title="例 2.9: 同じテストでの @depends と @dataProvider の組み合わせ">例 2.9</a>
      を参照ください。
    </p><div class="example"><a id="writing-tests-for-phpunit.data-providers.examples.DependencyAndDataProviderCombo.php"></a><p class="title"><strong>例 2.9: 同じテストでの @depends と @dataProvider の組み合わせ</strong></p><div class="example-contents"><pre class="programlisting">&lt;?php
class DependencyAndDataProviderComboTest extends PHPUnit_Framework_TestCase
{
    public function provider()
    {
        return array(array('provider1'), array('provider2'));
    }

    public function testProducerFirst()
    {
        $this-&gt;assertTrue(true);
        return 'first';
    }

    public function testProducerSecond()
    {
        $this-&gt;assertTrue(true);
        return 'second';
    }

    /**
     * @depends testProducerFirst
     * @depends testProducerSecond
     * @dataProvider provider
     */
    public function testConsumer()
    {
        $this-&gt;assertEquals(
            array('provider1', 'first', 'second'),
            func_get_args()
        );
    }
}
?&gt;</pre><pre class="screen"><strong class="userinput"><code>phpunit --verbose DependencyAndDataProviderComboTest</code></strong>
PHPUnit 5.3.0 by Sebastian Bergmann and contributors.

...F

Time: 0 seconds, Memory: 3.50Mb

There was 1 failure:

1) DependencyAndDataProviderComboTest::testConsumer with data set #1 ('provider2')
Failed asserting that two arrays are equal.
--- Expected
+++ Actual
@@ @@
Array (
-    0 =&gt; 'provider1'
+    0 =&gt; 'provider2'
1 =&gt; 'first'
2 =&gt; 'second'
)

/home/sb/DependencyAndDataProviderComboTest.php:31

FAILURES!
Tests: 4, Assertions: 4, Failures: 1.
</pre></div></div><br class="example-break"></br><div class="alert alert-info"><h3 class="title">注記</h3><p>
        <a id="idp355232" class="indexterm"></a>
        <a id="idp355808" class="indexterm"></a>
        <a id="idp356384" class="indexterm"></a>

        あるテストがデータプロバイダを使う別のテストに依存している場合、
        別のテストで少なくともひとつのデータセットに対するテストが成功すれば
        そのテストも実行されます。
        データプロバイダを使ったテストの結果をそのテストに注入することはできません。
      </p></div><div class="alert alert-info"><h3 class="title">注記</h3><p>
        <a id="idp358240" class="indexterm"></a>
        <a id="idp358816" class="indexterm"></a>
        <a id="idp359392" class="indexterm"></a>

        すべてのデータプロバイダを実行してから、
        静的メソッド <code class="literal">setUpBeforeClass</code>
        や <code class="literal">setUp</code> メソッドの最初の呼び出しが発生します。そのため、
        これらのメソッドで作った変数にデータプロバイダ内からアクセスすることはできません。
        そうなっている理由は、PHPUnit がテストの総数を算出できるようにするためです。
      </p></div></div><div class="section"><div class="titlepage"><div><div><h2 class="title" style="clear: both"><a id="writing-tests-for-phpunit.exceptions"></a>例外のテスト</h2></div></div></div><p>
      <a id="idp363056" class="indexterm"></a>
      <a id="idp363632" class="indexterm"></a>

      <a class="xref" href="writing-tests-for-phpunit.html#writing-tests-for-phpunit.exceptions.examples.ExceptionTest.php" title="例 2.10: expectException() メソッドの使用法">例 2.10</a>
      は、テストするコード内で例外がスローされたかどうかを
      <code class="literal">expectException()</code> メソッドを使用して調べる方法を示すものです。
    </p><div class="example"><a id="writing-tests-for-phpunit.exceptions.examples.ExceptionTest.php"></a><p class="title"><strong>例 2.10: expectException() メソッドの使用法</strong></p><div class="example-contents"><pre class="programlisting">&lt;?php
class ExceptionTest extends PHPUnit_Framework_TestCase
{
    public function testException()
    {
        $this-&gt;expectException(InvalidArgumentException::class);
    }
}
?&gt;</pre><pre class="screen"><strong class="userinput"><code>phpunit ExceptionTest</code></strong>
PHPUnit 5.3.0 by Sebastian Bergmann and contributors.

F

Time: 0 seconds, Memory: 4.75Mb

There was 1 failure:

1) ExceptionTest::testException
Expected exception InvalidArgumentException

FAILURES!
Tests: 1, Assertions: 1, Failures: 1.</pre></div></div><br class="example-break"></br><p>
      <a id="idp368672" class="indexterm"></a>
      <a id="idp369248" class="indexterm"></a>
      <a id="idp369840" class="indexterm"></a>

      さらに、<code class="literal">expectException()</code>、
      <code class="literal">expectExceptionCode()</code>、
      <code class="literal">expectExceptionMessage()</code>、
      <code class="literal">expectExceptionMessageRegExp()</code> といったメソッドで、
      テスト対象のコードで発生するであろう例外をテストできます。
    </p><p>
      <a id="idp372816" class="indexterm"></a>
      <a id="idp373392" class="indexterm"></a>

      別の方法として、<code class="literal">@expectedException</code>、
      <code class="literal">@expectedExceptionCode</code>、
      <code class="literal">@expectedExceptionMessage</code>、
      <code class="literal">@expectedExceptionMessageRegExp</code> といったアノテーションでも、
      テスト対象のコードで発生するであろう例外をテストできます。
      <a class="xref" href="writing-tests-for-phpunit.html#writing-tests-for-phpunit.exceptions.examples.ExceptionTest2.php" title="例 2.11: @expectedException アノテーションの使用法">例 2.11</a>
      に例を示します。
    </p><div class="example"><a id="writing-tests-for-phpunit.exceptions.examples.ExceptionTest2.php"></a><p class="title"><strong>例 2.11: @expectedException アノテーションの使用法</strong></p><div class="example-contents"><pre class="programlisting">&lt;?php
class ExceptionTest extends PHPUnit_Framework_TestCase
{
    /**
     * @expectedException InvalidArgumentException
     */
    public function testException()
    {
    }
}
?&gt;</pre><pre class="screen"><strong class="userinput"><code>phpunit ExceptionTest</code></strong>
PHPUnit 5.3.0 by Sebastian Bergmann and contributors.

F

Time: 0 seconds, Memory: 4.75Mb

There was 1 failure:

1) ExceptionTest::testException
Expected exception InvalidArgumentException

FAILURES!
Tests: 1, Assertions: 1, Failures: 1.</pre></div></div><br class="example-break"></br></div><div class="section"><div class="titlepage"><div><div><h2 class="title" style="clear: both"><a id="writing-tests-for-phpunit.errors"></a>PHP のエラーのテスト</h2></div></div></div><p>
      <a id="idp381024" class="indexterm"></a>
      <a id="idp381600" class="indexterm"></a>
      <a id="idp382176" class="indexterm"></a>
      <a id="idp382752" class="indexterm"></a>
      <a id="idp383328" class="indexterm"></a>

      デフォルトでは、PHPUnit はテストの実行中に発生した PHP のエラーや警告そして notice
      を例外に変換します。これらの例外を用いて、たとえば
      <a class="xref" href="writing-tests-for-phpunit.html#writing-tests-for-phpunit.exceptions.examples.ErrorTest.php" title="例 2.12: @expectedException を用いた、PHP エラーが発生することのテスト">例 2.12</a>
      のように PHP のエラーが発生することをテストできます。
    </p><div class="alert alert-info"><h3 class="title">注記</h3><p>
        PHP の実行時設定 <code class="literal">error_reporting</code> を使うと、
        PHPUnit がどのエラーを例外に変換するのかを制限できます。
        この機能に関して何か問題がでた場合は、PHP の設定を見直し、
        調べたいと思っているエラーを抑制するようになっていないかどうか確認しましょう。
      </p></div><div class="example"><a id="writing-tests-for-phpunit.exceptions.examples.ErrorTest.php"></a><p class="title"><strong>例 2.12: @expectedException を用いた、PHP エラーが発生することのテスト</strong></p><div class="example-contents"><pre class="programlisting">&lt;?php
class ExpectedErrorTest extends PHPUnit_Framework_TestCase
{
    /**
     * @expectedException PHPUnit_Framework_Error
     */
    public function testFailingInclude()
    {
        include 'not_existing_file.php';
    }
}
?&gt;</pre><pre class="screen"><strong class="userinput"><code>phpunit -d error_reporting=2 ExpectedErrorTest</code></strong>
PHPUnit 5.3.0 by Sebastian Bergmann and contributors.

.

Time: 0 seconds

OK (1 test, 1 assertion)</pre></div></div><br class="example-break"></br><p>
      <a id="idp389696" class="indexterm"></a>
      <a id="idp390288" class="indexterm"></a>

      <code class="literal">PHPUnit_Framework_Error_Notice</code> および
      <code class="literal">PHPUnit_Framework_Error_Warning</code> は、
      それぞれ PHP の notice と警告に対応します。
    </p><div class="alert alert-info"><h3 class="title">注記</h3><p>
        例外をテストするときには可能な限り限定的にしなければいけません。
        あまりに一般化されすぎたクラスをテストすると、予期せぬ副作用を引き起こしかねません。
        というわけで、
        <code class="literal">@expectedException</code> や
        <code class="literal">setExpectedException()</code>
        を使った <code class="literal">Exception</code>
        クラスのテストはできないようにしました。
      </p></div><p>
        エラーを引き起こすような PHP の関数、たとえば <code class="literal">fopen</code>
        などに依存するテストを行うときには、テスト中にエラーを抑制できれば便利なことがあります。
        そうすれば、notice のせいで
        <code class="literal">PHPUnit_Framework_Error_Notice</code>
        が出てしまうことなく、返り値だけをチェックできるようになります。
        </p><div class="example"><a id="writing-tests-for-phpunit.exceptions.examples.TriggerErrorReturnValue.php"></a><p class="title"><strong>例 2.13: PHP のエラーが発生するコードの返り値のテスト</strong></p><div class="example-contents"><pre class="programlisting">&lt;?php
class ErrorSuppressionTest extends PHPUnit_Framework_TestCase
{
    public function testFileWriting() {
        $writer = new FileWriter;
        $this-&gt;assertFalse(@$writer-&gt;write('/is-not-writeable/file', 'stuff'));
    }
}
class FileWriter
{
    public function write($file, $content) {
        $file = fopen($file, 'w');
        if($file == false) {
            return false;
        }
        // ...
    }
}

?&gt;</pre><pre class="screen"><strong class="userinput"><code>phpunit ErrorSuppressionTest</code></strong>
PHPUnit 5.3.0 by Sebastian Bergmann and contributors.

.

Time: 1 seconds, Memory: 5.25Mb

OK (1 test, 1 assertion)</pre></div></div><p><br class="example-break"></br>
    もしエラーを抑制しなければ、このテストは失敗して
    <code class="literal">fopen(/is-not-writeable/file): failed to open stream:
    No such file or directory</code> となります。

    </p></div><div class="section"><div class="titlepage"><div><div><h2 class="title" style="clear: both"><a id="writing-tests-for-phpunit.output"></a>出力内容のテスト</h2></div></div></div><p>
      メソッドの実行結果を確かめる方法として、(<code class="literal">echo</code> や
      <code class="literal">print</code> などによる)
      出力が期待通りのものかを調べたいこともあるでしょう。
      <code class="literal">PHPUnit_Framework_TestCase</code> クラスは、PHP の
      <a class="ulink" href="http://www.php.net/manual/ja/ref.outcontrol.php" target="_top">
      出力バッファリング</a> 機能を使用してこの仕組みを提供します。
    </p><p>
      <a class="xref" href="writing-tests-for-phpunit.html#writing-tests-for-phpunit.output.examples.OutputTest.php" title="例 2.14: 関数やメソッドの出力内容のテスト">例 2.14</a>
      では、期待する出力内容を <code class="literal">expectOutputString()</code>
      メソッドで設定する方法を示します。
      期待通りの出力が得られなかった場合は、そのテストは失敗という扱いになります。
    </p><div class="example"><a id="writing-tests-for-phpunit.output.examples.OutputTest.php"></a><p class="title"><strong>例 2.14: 関数やメソッドの出力内容のテスト</strong></p><div class="example-contents"><pre class="programlisting">&lt;?php
class OutputTest extends PHPUnit_Framework_TestCase
{
    public function testExpectFooActualFoo()
    {
        $this-&gt;expectOutputString('foo');
        print 'foo';
    }

    public function testExpectBarActualBaz()
    {
        $this-&gt;expectOutputString('bar');
        print 'baz';
    }
}
?&gt;</pre><pre class="screen"><strong class="userinput"><code>phpunit OutputTest</code></strong>
PHPUnit 5.3.0 by Sebastian Bergmann and contributors.

.F

Time: 0 seconds, Memory: 5.75Mb

There was 1 failure:

1) OutputTest::testExpectBarActualBaz
Failed asserting that two strings are equal.
--- Expected
+++ Actual
@@ @@
-'bar'
+'baz'


FAILURES!
Tests: 2, Assertions: 2, Failures: 1.</pre></div></div><br class="example-break"></br><p>
      <a class="xref" href="writing-tests-for-phpunit.html#writing-tests-for-phpunit.output.tables.api" title="表2.1 テストの出力用のメソッド">表 2.1</a> は、
      テストの出力用に提供するメソッドをまとめたものです。
    </p><div class="table"><a id="writing-tests-for-phpunit.output.tables.api"></a><p class="title"><strong>表2.1 テストの出力用のメソッド</strong></p><div class="table-contents"><table class="table" summary="テストの出力用のメソッド" border="1"><colgroup><col></col><col></col></colgroup><thead><tr><th align="left">メソッド</th><th align="left">意味</th></tr></thead><tbody><tr><td align="left"><code class="literal" style="white-space: normal" id="expectOutputRegex">void expectOutputRegex(string $regularExpression)</code></td><td align="left">出力が正規表現 <code class="literal">$regularExpression</code> にマッチするであろうという予測を設定します。</td></tr><tr><td align="left"><code class="literal" style="white-space: normal" id="expectOutputString">void expectOutputString(string $expectedString)</code></td><td align="left">出力が文字列 <code class="literal">$expectedString</code> と等しくなるであろうという予測を設定します。</td></tr><tr><td align="left"><code class="literal" style="white-space: normal" id="setOutputCallback">bool setOutputCallback(callable $callback)</code></td><td align="left">たとえば出力時の正規化などに使用するコールバック関数を設定します。</td></tr></tbody></table></div></div><br class="table-break"></br><div class="alert alert-info"><h3 class="title">注記</h3><p>
        strict モードでは、出力を発生させるテストは失敗します。
      </p></div></div><div class="section"><div class="titlepage"><div><div><h2 class="title" style="clear: both"><a id="writing-tests-for-phpunit.error-output"></a>エラー出力</h2></div></div></div><p>
      テストが失敗した場合、PHPUnit は、状況を可能な限り詳細に報告します。
      これが、何が問題だったのかを調べるのに役立つでしょう。
    </p><div class="example"><a id="writing-tests-for-phpunit.error-output.examples.ArrayDiffTest.php"></a><p class="title"><strong>例 2.15: 配列の比較に失敗したときのエラー出力</strong></p><div class="example-contents"><pre class="programlisting">&lt;?php
class ArrayDiffTest extends PHPUnit_Framework_TestCase
{
    public function testEquality() {
        $this-&gt;assertEquals(
            array(1,2,3 ,4,5,6),
            array(1,2,33,4,5,6)
        );
    }
}
?&gt;</pre><pre class="screen"><strong class="userinput"><code>phpunit ArrayDiffTest</code></strong>
PHPUnit 5.3.0 by Sebastian Bergmann and contributors.

F

Time: 0 seconds, Memory: 5.25Mb

There was 1 failure:

1) ArrayDiffTest::testEquality
Failed asserting that two arrays are equal.
--- Expected
+++ Actual
@@ @@
 Array (
     0 =&gt; 1
     1 =&gt; 2
-    2 =&gt; 3
+    2 =&gt; 33
     3 =&gt; 4
     4 =&gt; 5
     5 =&gt; 6
 )

/home/sb/ArrayDiffTest.php:7

FAILURES!
Tests: 1, Assertions: 1, Failures: 1.</pre></div></div><br class="example-break"></br><p>
      この例では配列の要素のうちひとつだけが異なっています。
      それ以外の値も表示することで、どこが悪かったのかをわかりやすくしています。
    </p><p>
      出力が長すぎる場合は PHPUnit が出力を分割し、違っている部分の前後数行だけを出力します。
    </p><div class="example"><a id="writing-tests-for-phpunit.error-output.examples.LongArrayDiffTest.php"></a><p class="title"><strong>例 2.16: 要素数の多い配列の比較に失敗したときのエラー出力</strong></p><div class="example-contents"><pre class="programlisting">&lt;?php
class LongArrayDiffTest extends PHPUnit_Framework_TestCase
{
    public function testEquality() {
        $this-&gt;assertEquals(
            array(0,0,0,0,0,0,0,0,0,0,0,0,1,2,3 ,4,5,6),
            array(0,0,0,0,0,0,0,0,0,0,0,0,1,2,33,4,5,6)
        );
    }
}
?&gt;</pre><pre class="screen"><strong class="userinput"><code>phpunit LongArrayDiffTest</code></strong>
PHPUnit 5.3.0 by Sebastian Bergmann and contributors.

F

Time: 0 seconds, Memory: 5.25Mb

There was 1 failure:

1) LongArrayDiffTest::testEquality
Failed asserting that two arrays are equal.
--- Expected
+++ Actual
@@ @@
     13 =&gt; 2
-    14 =&gt; 3
+    14 =&gt; 33
     15 =&gt; 4
     16 =&gt; 5
     17 =&gt; 6
 )


/home/sb/LongArrayDiffTest.php:7

FAILURES!
Tests: 1, Assertions: 1, Failures: 1.</pre></div></div><br class="example-break"></br><div class="section"><div class="titlepage"><div><div><h3 class="title"><a id="writing-tests-for-phpunit.error-output.edge-cases"></a>エッジケース</h3></div></div></div><p>
        比較に失敗したときに、PHPUnit は入力値をテキスト形式にしてこれを比較します。
        この実装が原因で、実際の違う箇所よりも多くの問題を報告してしまうことがあります。
      </p><p>
        この問題が発生するのは、
        assertEquals などの「緩い」比較の関数を、配列やオブジェクトに対して使った場合だけです。
      </p><div class="example"><a id="writing-tests-for-phpunit.error-output.edge-cases.examples.ArrayWeakComparisonTest.php"></a><p class="title"><strong>例 2.17: 緩い比較を使った場合の diff の生成のエッジケース</strong></p><div class="example-contents"><pre class="programlisting">&lt;?php
class ArrayWeakComparisonTest extends PHPUnit_Framework_TestCase
{
    public function testEquality() {
        $this-&gt;assertEquals(
            array(1  ,2,3 ,4,5,6),
            array('1',2,33,4,5,6)
        );
    }
}
?&gt;</pre><pre class="screen"><strong class="userinput"><code>phpunit ArrayWeakComparisonTest</code></strong>
PHPUnit 5.3.0 by Sebastian Bergmann and contributors.

F

Time: 0 seconds, Memory: 5.25Mb

There was 1 failure:

1) ArrayWeakComparisonTest::testEquality
Failed asserting that two arrays are equal.
--- Expected
+++ Actual
@@ @@
 Array (
-    0 =&gt; 1
+    0 =&gt; '1'
     1 =&gt; 2
-    2 =&gt; 3
+    2 =&gt; 33
     3 =&gt; 4
     4 =&gt; 5
     5 =&gt; 6
 )


/home/sb/ArrayWeakComparisonTest.php:7

FAILURES!
Tests: 1, Assertions: 1, Failures: 1.</pre></div></div><br class="example-break"></br><p>
        この例では、最初のインデックスの
        <code class="literal">1</code> と <code class="literal">'1'</code>
        がエラー報告されていますが、assertEquals ではこれらを等しいとみなしているはずです。
      </p></div></div></div>
     <div class="row">
      <div class="col-md-1 pull-left prev-nav"><a accesskey="p" href="installation.html">戻る</a></div>
      <div class="col-md-1 pull-right next-nav"><a accesskey="n" href="textui.html">次へ</a></div>
     </div>



    </div>
   </div>
   <hr/>
   <footer>
    <p><a href="appendixes.copyright.html">Copyright</a> &copy; 2005-2016 <a href="http://sebastian-bergmann.de/">Sebastian Bergmann</a>.</p>
   </footer>
  </div>
  <script src="js/jquery.js"></script>
  <script src="js/bootstrap.min.js"></script>
  <script src="js/highlight.pack.js"></script>
  <script type="text/javascript">
  $(document).ready(function() { $('pre.programlisting').each(function(i, e) {hljs.highlightBlock(e)}); });
  </script>
 </body>
</html>
